Dockerfile

```diff
@@ -1,44 +1,200 @@
+
+###############################################################################
+#                          FULL TEST STAGE
+###############################################################################
+FROM full_build AS test_full
+
+RUN python -m pip install --no-cache-dir -e "$GEOIPS_PACKAGES_DIR/geoips/[doc,test]" \
+    && echo "import coverage; coverage.process_startup()" > ~/.local/lib/python3.10/site-packages/coverage.pth
+# See https://coverage.readthedocs.io/en/coverage-5.1/subprocess.html#measuring-sub-processes
+# for more info on how this helps measure coverage of subprocess-based tests
+
+ENTRYPOINT ["pytest"]
+
+###############################################################################
+#                          SYSTEM TEST STAGE
+###############################################################################
+FROM system_build AS test_system
+
+RUN python -m pip install --no-cache-dir -e "$GEOIPS_PACKAGES_DIR/geoips/[doc,test]" \
+    && echo "import coverage; coverage.process_startup()" > ~/.local/lib/python3.10/site-packages/coverage.pth
+# See https://coverage.readthedocs.io/en/coverage-5.1/subprocess.html#measuring-sub-processes
+# for more info on how this helps measure coverage of subprocess-based tests
+
+ENTRYPOINT ["pytest"]
+
+###############################################################################
+#                           DEV STAGE
+###############################################################################
+FROM test_system AS dev
+
+# Install lint, debug, etc. on top of full test
+RUN python -m pip install --no-cache-dir -e "$GEOIPS_PACKAGES_DIR/geoips/[doc,test,lint,debug]"
+
+###############################################################################
+#                              PRODUCTION STAGE
+###############################################################################
+FROM build AS production
+
+USER root
+
+# Remove unnecessary files for smaller production image
+RUN cd "$GEOIPS_PACKAGES_DIR/geoips" \
+    && rm -rf \
+       CHANGELOG_TEMPLATE.rst \
+       environments \
+       .github \
+       pyproject.toml \
+       tests \
+       CODE_OF_CONDUCT.md \
+       Dockerfile \
+       .flake8 \
+       .gitignore \
+       pytest.ini \
+       update_this_release_note \
+       bandit.yml \
+       COMMIT_MESSAGE_TEMPLATE.md \
+       .dockerignore \
+       interface_notes.md \
+       README.md \
+       CHANGELOG.rst \
+       .config \
+       docs \
+       setup \
+       requirements.txt \
+    && apt-get remove -y \
+       git \
+       make \
+       g++ \
+       wget \
+       gfortran \
+    && apt-get autoremove -y \
+    && rm -rf /var/lib/apt/lists/*
+
+# Switch back to non-root for runtime
+USER ${USER}
+WORKDIR ${GEOIPS_PACKAGES_DIR}/geoips
+
+# By default, provide a simple help CMD and a direct ENTRYPOINT
+ENTRYPOINT ["geoips"]
+CMD ["geoips", "--help"]
```

Comment:
I think that we may want to consider having two completely different paths through this dockerfile. One that is intended for dev/test and one that is intended for production. We may also need some testing for production to ensure that things work when installed without `-e`. The `production` stage should be installed without `-e`.

This can be done in a follow-on issue but please create the issue and link it here.